src/bootman/kernel.c

```diff
@@ -592,36 +592,26 @@ bool boot_manager_install_kernel_internal(const BootManager *manager, const Kern
         autofree(char) *base_path = NULL;
         autofree(char) *initrd_target = NULL;
         const char *initrd_source = NULL;
-        bool is_uefi = false;
-        autofree(char) *efi_boot_dir = NULL;
+        int is_uefi = (manager->bootloader->get_capabilities(manager) & BOOTLOADER_CAP_UEFI);
+        autofree(char) *efi_boot_dir =
+            is_uefi ? manager->bootloader->get_kernel_dst(manager) : NULL;
 
         assert(manager != NULL);
         assert(kernel != NULL);
 
+        if (is_uefi && !efi_boot_dir)
+                return false;
+
         /* Boot path */
         base_path = boot_manager_get_boot_dir((BootManager *)manager);
         OOM_CHECK_RET(base_path, false);
 
-        /* Determine if UEFI is in use */
-        if ((manager->bootloader->get_capabilities(manager) & BOOTLOADER_CAP_UEFI) ==
-            BOOTLOADER_CAP_UEFI) {
-                is_uefi = true;
-                efi_boot_dir = nc_build_case_correct_path(base_path, "EFI", KERNEL_NAMESPACE, NULL);
-        }
-
-        /* For UEFI kernels we namespace into /EFI/$NEEDLE, i.e. /EFI/org.clearlinux */
-        if (is_uefi) {
-                /* Ensure namespace directory exists */
-                if (!nc_mkdir_p(efi_boot_dir, 00755)) {
-                        LOG_FATAL("Failed to create namespace directory: %s %s",
-                                  efi_boot_dir,
-                                  strerror(errno));
-                        return false;
-                }
-                kfile_target = string_printf("%s/%s", efi_boot_dir, kernel->target.path);
-        } else {
-                kfile_target = string_printf("%s/%s", base_path, kernel->target.legacy_path);
-        }
+        /* for UEFI, the kernel location is prefixed with efi_boot_dir which is
+         * guaranteed to start with '/' since it's its absolute path on ESP. */
+        kfile_target = string_printf("%s%s/%s",
```

Comment:
This is less clean/easy to follow than before.

Reply:
the comment explains the why, does it not?